File: 백준/Gold/14719. 빗물/빗물.js
const fs = require("fs");
const stdin = (
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString()
    : fs.readFileSync("./input.txt").toString()
).split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const [W, H] = input()
  .trim()
  .split(" ")
  .map((v) => parseInt(v));

const blocks = input()
  .trim()
  .split(" ")
  .map((v) => parseInt(v));

let answer = 0;

for (let i = 1; i < H - 1; i++) {
  const left = Math.max(...blocks.slice(0, i));
  const right = Math.max(...blocks.slice(i + 1));

  const minVal = Math.min(left, right);

  if (blocks[i] < minVal) answer += minVal - blocks[i];
}

console.log(answer);

File: 백준/Gold/14719. 빗물/빗물.java
import java.io.*;
import java.util.*;

public class Main {

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		StringTokenizer st = new StringTokenizer(br.readLine(), " ");

		int H = Integer.parseInt(st.nextToken());
		int W = Integer.parseInt(st.nextToken());

		st = new StringTokenizer(br.readLine(), " ");
		int[] map = new int[W];

		for (int w = 0; w < W; w++) {
			map[w] = Integer.parseInt(st.nextToken());
		}

		int answer = 0;

		for (int w = 1; w < W - 1; w++) {
			int left = 0, right = 0;

			for (int l = 0; l < w; l++) {
				left = Math.max(map[l], left);
			}

			for (int r = w + 1; r < W; r++) {
				right = Math.max(map[r], right);
			}

			if (map[w] < left && map[w] < right)
				answer += Math.min(left, right) - map[w];

		}

		System.out.println(answer);

		br.close();
	}

}
